tests/test_make.py

```diff
@@ -632,6 +641,154 @@ def test_sugar_callable(self):
+
+    def test_uninitialized_singleton(self):
+        """
+        Test that singleton raises ValueError if non-defaulted attr not defined
+        """
+
+        @attr.singleton()
+        @attr.s(frozen=True)
+        class C(object):
+            x = attr.ib()
+
+        with pytest.raises(ValueError):
+            C()
+
+    def test_singleton_init_only_once(self):
+        """
+        Test that a singleton's __init__ only runs *once*
+        """
+
+        @attr.singleton
+        @attr.s(frozen=True)
+        class C(object):
+            x = attr.ib(default=3)
+            ran = 0
+
+            def __attrs_post_init__(self, *args, **kwargs):
+                C.ran += 1
+
+        C()
+        C()
+        assert 1 == C.ran
+
+    def test_singleton(self):
+        """
+        Ensure that classes can be made singletons
+        """
+
+        @attr.singleton
+        @attr.s(frozen=True)
+        class C(object):
+            x = attr.ib(default=3)
+
+        a = C(4)
+        b = C(x=4)
+        c = C()
+        assert a is b
+        assert a is not c
+        assert a.x is b.x
+        assert a.x != c.x
+
+    def test_singleton_from_attrs(self):
+        """
+        Ensure that classes can be made singletons from the attrs decorator
+        """
+
+        @attr.s(singleton=True)
+        class C(object):
+            x = attr.ib(default=3)
+
+        a = C(4)
+        b = C(x=4)
+        c = C()
+        assert a is b
+        assert a is not c
+        assert a.x is b.x
+        assert a.x != c.x
+
+    def test_singleton_collection_namedtuple(self):
+        """
+        Ensure that namedtuple can be made singletons
+        """
+
+        @attr.singleton
+        class C(namedtuple("_C", ["x"])):
+            pass
+
+        a = C(4)
+        b = C(x=4)
+        c = C(3)
+        assert a is b
+        assert a is not c
+        assert a.x is b.x
+        assert a.x != c.x
+
+    @pytest.mark.skipif(no_typing, reason="Typing is not installed to test")
+    def test_singleton_typing_namedtuple(self):
+        """
+        Ensure that namedtuple can be made singletons
+        """
+
+        @attr.singleton
+        class C(NamedTuple("_C", [("x", int)])):
+            pass
+
+        a = C(4)
+        b = C(x=4)
+        c = C(3)
+        assert a is b
+        assert a is not c
+        assert a.x is b.x
+        assert a.x != c.x
+
+    @pytest.mark.skipif(not PY_GE_36, reason="Pretty NamedTuple is only >=3.6")
+    def test_singleton_typing_namedtuple_36(self):
+        """
+        Ensure that namedtuple can be made singletons
+        """
+        exec(
+            "@attr.singleton\n"
+            "class C(NamedTuple):\n"
+            "    x: int = 3\n\n"
+            "a = C(4)\n"
+            "b = C(x=4)\n"
+            "c = C()\n"
+            "assert a is b\n"
+            "assert a is not c\n"
+            "assert a.x is b.x\n"
+            "assert a.x != c.x"
+        )  # avoid python2 syntax error
```

Comment:
There are probably better ways to format this test, but I wanted to avoid a python2 syntax error. Again, open to suggestions for improvement